### ISSUE TITLE
kube-proxy trying to talk to 10.0.0.1

### ISSUE THREAD

**Issue by colemickens** [kubernetes/kubernetes-anywhere#232]
User spun up cluster. Things weren't working because kube-proxy was in a bad state.

`kube-proxy` pods were up and running... but they were in a failure loop:

```shell
kubectl logs --namespace kube-system kube-proxy-bodsy   
Flag --resource-container has been deprecated, This feature will be removed in a later release.
E0927 02:46:21.189588       1 server.go:369] Can't get Node "ralphk8s-node-3", assuming iptables proxy, err: Get https://10.0.0.1:443/api/v1/nodes/ralphk8s-node-3: dial tcp 10.0.0.1:443: i/o timeout
I0927 02:46:21.191739       1 server.go:203] Using iptables Proxier.
W0927 02:46:51.192428       1 server.go:417] Failed to retrieve node info: Get https://10.0.0.1:443/api/v1/nodes/ralphk8s-node-3: dial tcp 10.0.0.1:443: i/o timeout
W0927 02:46:51.192724       1 proxier.go:226] invalid nodeIP, initialize kube-proxy with 127.0.0.1 as nodeIP
I0927 02:46:51.192793       1 server.go:215] Tearing down userspace rules.
I0927 02:46:51.203663       1 conntrack.go:40] Setting nf_conntrack_max to 32768
I0927 02:46:51.204145       1 conntrack.go:57] Setting conntrack hashsize to 8192
I0927 02:46:51.212064       1 conntrack.go:62] Setting nf_conntrack_tcp_timeout_established to 86400
E0927 02:47:21.205309       1 reflector.go:203] pkg/proxy/config/api.go:30: Failed to list *api.Service: Get https://10.0.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 02:47:21.212587       1 event.go:208] Unable to write event: 'Post https://10.0.0.1:443/api/v1/namespaces/default/events: dial tcp 10.0.0.1:443: i/o timeout' (may retry after sleeping)
E0927 02:47:21.212806       1 reflector.go:203] pkg/proxy/config/api.go:33: Failed to list *api.Endpoints: Get https://10.0.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 02:47:52.206429       1 reflector.go:203] pkg/proxy/config/api.go:30: Failed to list *api.Service: Get https://10.0.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 02:47:52.215875       1 reflector.go:203] pkg/proxy/config/api.go:33: Failed to list *api.Endpoints: Get https://10.0.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 02:47:55.750822       1 event.go:208] Unable to write event: 'Post https://10.0.0.1:443/api/v1/namespaces/default/events: dial tcp 10.0.0.1:443: i/o timeout' (may retry after sleeping)
E0927 02:48:23.207747       1 reflector.go:203] pkg/proxy/config/api.go:30: Failed to list *api.Service: Get https://10.0.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 02:48:23.216786       1 reflector.go:203] pkg/proxy/config/api.go:33: Failed to list *api.Endpoints: Get https://10.0.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 02:48:35.751387       1 event.go:208] Unable to write event: 'Post https://10.0.0.1:443/api/v1/namespaces/default/events: dial tcp 10.0.0.1:443: i/o timeout' (may retry after sleeping)
E0927 02:48:54.208682       1 reflector.go:203] pkg/proxy/config/api.go:30: Failed to list *api.Service: Get https://10.0.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 02:48:54.217860       1 reflector.go:203] pkg/proxy/config/api.go:33: Failed to list *api.Endpoints: Get https://10.0.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
```

@mikedanese Have you ever seen this? I can't come up with a single good reason this would happen. I changed the the non-beta hyperkube image for kube-proxy and killed the `kube-proxy` pods and it started working. I then reverted back to the original image and checked and it was still working. I have no idea how it got into that bad state...

The `kubeconfig` had to have been there (Ignition finished before apiserver can possibly get up)... so why would `kube-proxy` be trying to use the service address... it could never work...

CC: @squillace

**Comment by colemickens** [comment]
Only thing I see is that this is using `Standard_D1_v2` whereas I tend to use `Standard_D2_v2`. Still, I don't think any kind of race in `kubernetes-anywhere` explains this behavior:

1. `kubeconfig` is guaranteed to be in place before `kube-apiserver` comes up... let alone `kube-proxy`.

2. the `kube-proxy` yaml clearly passes the `--kubeconfig` flag to `kube-proxy`.

I'm trying to figure out if there was a bug in `hyperkube-amd:v1.4.0-beta.3` which caused `kube-proxy` to sometimes choose the wrong endpoint for apiserver...

I've sent PR #233 to try and address this in case there was a point-in-time bug in `kube-proxy`.

**Comment by edevil** [comment]
I have the same problem. Just spun up a test cluster to experiment with the Azure cloud provider and pods cannot connect to the kubernetes api server on 10.0.0.1:
```
kubectl logs kubernetes-dashboard-1655269645-hp4z6 --namespace=kube-system
Starting HTTP server on port 9090
Creating API server client for https://10.0.0.1:443
Error while initializing connection to Kubernetes apiserver. This most likely means that the cluster is misconfigured (e.g., it has invalid apiserver certificates or service accounts configuration) or the --apiserver-host param points to a server that does not exist. Reason: Get https://10.0.0.1:443/version: dial tcp 10.0.0.1:443: i/o timeout
```

```
 kubectl logs kubernetes-dashboard-1655269645-hp4z6 --namespace=kube-system
Starting HTTP server on port 9090
Creating API server client for https://10.0.0.1:443
Error while initializing connection to Kubernetes apiserver. This most likely means that the cluster is misconfigured (e.g., it has invalid apiserver certificates or service accounts configuration) or the --apiserver-host param points to a server that does not exist. Reason: Get https://10.0.0.1:443/version: dial tcp 10.0.0.1:443: i/o timeout
[container]:/opt/kubernetes-anywhere/phase1/azure/.tmp> kubectl logs --namespace kube-system kube-proxy-0ewvv
Flag --resource-container has been deprecated, This feature will be removed in a later release.
E0927 13:47:13.295628       1 server.go:369] Can't get Node "kubeanywhere-node-0", assuming iptables proxy, err: Get https://10.0.0.1:443/api/v1/nodes/kubeanywhere-node-0: dial tcp 10.0.0.1:443: i/o timeout
I0927 13:47:13.299640       1 server.go:203] Using iptables Proxier.
W0927 13:47:43.300084       1 server.go:417] Failed to retrieve node info: Get https://10.0.0.1:443/api/v1/nodes/kubeanywhere-node-0: dial tcp 10.0.0.1:443: i/o timeout
W0927 13:47:43.300349       1 proxier.go:226] invalid nodeIP, initialize kube-proxy with 127.0.0.1 as nodeIP
I0927 13:47:43.300414       1 server.go:215] Tearing down userspace rules.
I0927 13:47:43.319682       1 conntrack.go:40] Setting nf_conntrack_max to 32768
I0927 13:47:43.320167       1 conntrack.go:57] Setting conntrack hashsize to 8192
I0927 13:47:43.320322       1 conntrack.go:62] Setting nf_conntrack_tcp_timeout_established to 86400
E0927 13:48:13.320958       1 reflector.go:203] pkg/proxy/config/api.go:30: Failed to list *api.Service: Get https://10.0.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 13:48:13.321267       1 event.go:208] Unable to write event: 'Post https://10.0.0.1:443/api/v1/namespaces/default/events: dial tcp 10.0.0.1:443: i/o timeout' (may retry after sleeping)
E0927 13:48:13.321435       1 reflector.go:203] pkg/proxy/config/api.go:33: Failed to list *api.Endpoints: Get https://10.0.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 13:48:44.321968       1 reflector.go:203] pkg/proxy/config/api.go:33: Failed to list *api.Endpoints: Get https://10.0.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 13:48:44.322219       1 reflector.go:203] pkg/proxy/config/api.go:30: Failed to list *api.Service: Get https://10.0.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.0.0.1:443: i/o timeout
E0927 13:48:50.500093       1 event.go:208] Unable to write event: 'Post https://10.0.0.1:443/api/v1/namespaces/default/events: dial tcp 10.0.0.1:443: i/o timeout' (may retry after sleeping)
```

**Comment by colemickens** [comment]
@edevil would you mind pulling the #233 PR and seeing if this still repros?

**Comment by edevil** [comment]
I tried deploying a cluster with #233 and did not have this problem, then tried again with master and the problem reappeared. This is not definitive proof, but the PR may fix this problem.

**Comment by ahmetalpbalkan** [comment]
+1 for @edevil, fixes the problem for me as well

**Comment by colemickens** [comment]
This seems to be resolved since #233 was merged.
